src/main/java/uk/nhs/hee/trainee/details/api/DspCredentialResource.java

```diff
@@ -0,0 +1,267 @@
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.client.HttpClientErrorException;
+import org.springframework.web.client.RestTemplate;
+import uk.nhs.hee.trainee.details.model.Placement;
+import uk.nhs.hee.trainee.details.model.ProgrammeMembership;
+import uk.nhs.hee.trainee.details.model.dsp.IssueTokenResponse;
+import uk.nhs.hee.trainee.details.model.dsp.ParResponse;
+import uk.nhs.hee.trainee.details.service.JwtService;
+import uk.nhs.hee.trainee.details.service.PlacementService;
+import uk.nhs.hee.trainee.details.service.ProgrammeMembershipService;
+
+@Slf4j
+@RestController
+@RequestMapping("/api/credential")
+public class DspCredentialResource {
+
+  private static final Integer CONNECT_TIMEOUT = 2000;
+  private static final Integer READ_TIMEOUT = 5000;
+  private static final String STATE = "someString"; //TODO consider this
+  private static final String TIS_ID_ATTRIBUTE = "custom:tisId";
+
+  private final ObjectMapper objectMapper;
+  private final PlacementService placementService;
+  private final ProgrammeMembershipService programmeMembershipService;
+  private final String clientId;
+  private final String clientSecret;
+  private final String redirectUri;
+  private final String parEndpoint;
+  private final String issueEndpoint;
+  private final String tokenEndpoint;
+  private final RestTemplateBuilder restTemplateBuilder;
+  private final JwtService jwtService;
+
+  /**
+   * Instantiate the DSP Credential Resource.
+   *
+   * @param placementService           the Placement service
+   * @param programmeMembershipService the Programme membership service
+   * @param objectMapper               the Object mapper
+   * @param clientId                   the DSP client ID
+   * @param clientSecret               the DSP client secret
+   * @param redirectUri                the DSP redirect URI
+   * @param parEndpoint                the DSP Pushed Authorization Request (PAR) endpoint
+   * @param authorizeEndpoint          the DSP credential Authorize Request endpoint
+   * @param tokenEndpoint              the DSP Token Request endpoint
+   * @param restTemplateBuilder        the REST template builder
+   * @param jwtService                 the JWT service
+   */
+  public DspCredentialResource(PlacementService placementService,
+                               ProgrammeMembershipService programmeMembershipService,
+                               ObjectMapper objectMapper,
+                               @Value("${dsp.client-id}") String clientId,
+                               @Value("${dsp.client-secret}") String clientSecret,
+                               @Value("${dsp.redirect-uri}") String redirectUri,
+                               @Value("${dsp.par-endpoint}") String parEndpoint,
+                               @Value("${dsp.authorize-endpoint}") String authorizeEndpoint,
+                               @Value("${dsp.token.issue-endpoint}") String tokenEndpoint,
```

Comment:
This is really messy, maybe look at `@ConfigurationProperties` as an alternative way to provide multiple related properties.

Reply:
I've tweaked this - is the change more or less what you had in mind? (or let me know if I've headed off at a tangent 😃)